src/docfx/lib/ProcessUtility.cs

```diff
@@ -98,5 +107,48 @@ public static Task<string> Execute(string fileName, string commandLineArgs, stri
 
             return tcs.Task;
         }
+
+        /// <summary>
+        /// Provide a process lock function based on locking file
+        /// </summary>
+        /// <param name="action">The action you want to lock</param>
+        /// <param name="lockPath">The lock file path, default is a file with GUID name</param>
+        /// <param name="retry">The retry count, default is 600 times</param>
+        /// <param name="retryTimeSpanInterval">The retry interval, default is 1 seconds</param>
+        /// <returns>The task status</returns>
+        public static async Task ProcessLock(Func<Task> action, string lockPath = null, int retry = 600, TimeSpan? retryTimeSpanInterval = null)
+        {
+            lockPath = lockPath ?? Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), "docfx", "lock", $"{Guid.NewGuid()}.lock");
+
+            Debug.Assert(!PathUtility.FilePathHasInvalidChars(lockPath));
+            Directory.CreateDirectory(Path.GetDirectoryName(lockPath));
+
+            var retryCount = 0;
+            var exceptions = new List<Exception>();
+            while (retryCount < retry)
+            {
+                try
+                {
+                    using (FileStream lockFile = new FileStream(lockPath, FileMode.OpenOrCreate, FileAccess.ReadWrite, FileShare.Delete))
+                    {
+                        await action();
+
+                        File.Delete(lockPath);
```

Comment:
put `Delete` in `finally` block

Reply:
It should be `FileOptions.DeleteOnClose` instead of `FileShare.Delete`. `FileOptions.DeleteOnClose` may not work on linux.

---
In reply to: [182346211](https://github.com/dotnet/docfx/pull/2651#discussion_r182346211) [](ancestors = 182346211)